torch/jit/__init__.py

```diff
@@ -621,29 +641,88 @@ def env(key):
         if lang is not None:
             self.define(lang, frame_id=3)
-        self.execution_engines = {}
         self.optimize = optimize
 
     def define(self, lang, rcb=None, frame_id=2):
         if not rcb:
             rcb = createResolutionCallback(frame_id)
-        self.cu.define(lang, rcb)
+        self.module._define(lang, rcb, False)
 
     def __getattr__(self, attr):
-        if attr not in self.execution_engines:
-            graph = self.cu.get_graph(attr)
-            self.execution_engines[attr] = torch._C.GraphExecutor(graph, self.optimize)
-        return self.execution_engines[attr]
+        return self.module._get_method(attr)
 
 
 def script(fn):
     rcb = createResolutionCallback()
-    ast = torch.jit.frontend.get_jit_ast(fn)
+    ast = get_jit_ast(fn)
     graph = _jit_script_compile(ast, rcb)
     return torch._C.GraphExecutor(graph, True)
 
+
+ScriptMethodStub = namedtuple('ScriptMethodStub', ('resolution_callback', 'ast'))
+
+
+def script_method(fn):
+    return ScriptMethodStub(createResolutionCallback(), get_jit_ast(fn))
+
+
+class ScriptMeta(type(torch._C.ScriptModule)):
+    # this has to inherit from pybind11's metaclass otherwise we get
+    # issues because ScriptModule inherits from torch._C.ScriptModule,
+    # a pybind11 type
+    def __init__(cls, name, bases, attrs):
+        # find all the script methods
+        methods = []
+        for k, v in sorted(attrs.items()):
+            if isinstance(v, ScriptMethodStub):
+                delattr(cls, k)
+                methods.append(v)
+        if len(methods) > 0 and '__init__' in attrs:
+            # after the user's __init__ register all the script methods
+            # with the module
+            original_init = cls.__init__
+
+            def init_then_register(self, *args, **kwargs):
+                original_init(self, *args, **kwargs)
+                for m in methods:
+                    self._create_method(m.ast, m.resolution_callback)
+
+            cls.__init__ = init_then_register
+        return super(ScriptMeta, cls).__init__(name, bases, attrs)
+
+
+class ScriptModule(with_metaclass(ScriptMeta, torch._C.ScriptModule)):
+
+    def __init__(self, optimize=True):
+        super(ScriptModule, self).__init__(optimize)
+
+    def __setattr__(self, name, value):
+        if isinstance(value, Parameter):
+            self._register_or_set_parameter(name, value)
+        elif isinstance(value, ScriptModule):
+            self._register_module(name, value)
+            # note: script modules are subclassed in python and the
+            # C++ script::Module class will not hold references to them
+            # to ensure that you always get the same python value here
+            # we store it as a native attribute _in addition to_
+            # registering it with the C++ script::Module
+            object.__setattr__(self, name, value)
+        else:
+            object.__setattr__(self, name, value)
```

Comment:
So is it true that now we really support a wide range of interactions with Python that we can't compile? In particular is accessing non-scripted members in scripted methods (think e.g. arbitrary python callables) and pure Python `nn.Modules` allowed? If not, where is it blocked/how does it fail?